Add clipmenu extension

Envl opened the pull request:
## Description
Show current clipboard content in Menu Bar and in Raycast Command

## Screencast
![Raycast 2023-11-24 at 21 23 08](https://github.com/raycast/extensions/assets/9925116/42538c6b-77e8-4a12-b2e4-c0bb87eababf)
![CleanShot 2023-11-24 at 21 30 47@2x](https://github.com/raycast/extensions/assets/9925116/f9670868-8e71-489a-8da0-7acf50940722)


## Changelog
### Added - 2023-12-08
- menubar item to open all clipboard history in raycast
- checkbox to configure whether show current clipboard content on menu bar or not
- dropdown to set text truncation limit


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented on extensions/clipmenu/src/utils.ts:
```diff
@@ -0,0 +1,7 @@
+export function iife(func: () => unknown) {
+  return func();
+}
+
+export async function iiafe(func: () => Promise<unknown>) {
+  return await func();
+}
```
```suggestion
export const truncateText = (text: string, length: number): string => {
  return text.length > length ? text.substring(0, length) + "..." : text;
};
```

pernielsentikaer commented on extensions/clipmenu/src/index.tsx:
```diff
@@ -0,0 +1,29 @@
+import { Clipboard, Icon, MenuBarExtra, updateCommandMetadata } from "@raycast/api";
+import { useCachedState } from "@raycast/utils";
+import { iiafe } from "./utils";
+
+export default function Command() {
+  const [history, setHistory] = useCachedState<string[]>("latest-clipboard-histories", [""]);
+
+  iiafe(async () => {
+    const _history = (
+      await Promise.all(
+        Array(6)
+          .fill(0)
+          .map((_, idx) => Clipboard.readText({ offset: idx })),
+      )
+    ).filter(Boolean) as string[];
+
+    updateCommandMetadata({ subtitle: _history[0] });
+    setHistory(_history);
+  });
+
+  return (
+    <MenuBarExtra icon={Icon.Clipboard} title={history[0]}>
+      <MenuBarExtra.Item title="Clipboard History" />
+      {history.map((text, i) => (
+        <MenuBarExtra.Item key={text + i} title={text} onAction={() => Clipboard.paste(text)} icon={Icon.Clipboard} />
+      ))}
+    </MenuBarExtra>
+  );
+}
```
```suggestion
import { Clipboard, Icon, MenuBarExtra, updateCommandMetadata } from "@raycast/api";
import { useCachedPromise } from "@raycast/utils";
import { truncateText } from "./utils";

export default function Command() {
  const { isLoading, data } = useCachedPromise(async () => {
    const _history = await Promise.all(
      Array(6)
        .fill(0)
        .map((_, idx) => Clipboard.readText({ offset: idx })),
    );

    updateCommandMetadata({ subtitle: _history[0] });
    return _history.filter(Boolean) as string[];
  });

  return (
    <MenuBarExtra isLoading={isLoading} icon={Icon.Clipboard} title={data ? truncateText(data[0], 50) : ""}>
      <MenuBarExtra.Section title="Clipboard History">
        {data &&
          data.map((text, i) => (
            <MenuBarExtra.Item
              key={text + i}
              title={truncateText(text, 50)}
              onAction={() => Clipboard.paste(text)}
              icon={Icon.Clipboard}
            />
          ))}
      </MenuBarExtra.Section>
    </MenuBarExtra>
  );
}
```

This might be cleaner, also, you can maybe add a preference to set how many items you want to show in the menubar (and maybe add if it should shown first element in menubar too (and length)

pernielsentikaer reviewed (changes requested):
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you

 

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

Envl commented:
Hi @pernielsentikaer, I've updated the code and added configurations.
regarding configuring "how many items to show", maybe it's not necessary as the documentations says we can get at most 6 clipboard records.

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
